cluster-autoscaler/utils/utils.go

```diff
@@ -54,3 +55,40 @@ func FilterOutNodes(nodes []*apiv1.Node, nodesToFilterOut []*apiv1.Node) []*apiv
 
 	return filtered
 }
+
+// PodSpecSemanticallyEqual returns true if two pod specs are similar after dropping
+// the fields we don't care about
+// Due to the generated suffixes, a strict DeepEquals check will fail and generate
+// an equivalence group per pod which is undesirable.
+// Projected volumes do not impact scheduling so we should ignore them
+func PodSpecSemanticallyEqual(p1 apiv1.PodSpec, p2 apiv1.PodSpec) bool {
+	p1Spec := sanitizeProjectedVolumesAndMounts(p1)
+	p2Spec := sanitizeProjectedVolumesAndMounts(p2)
+	return apiequality.Semantic.DeepEqual(p1Spec, p2Spec)
+}
+
+// sanitizeProjectedVolumesAndMounts returns a pod spec with projected volumes
+// and their mounts removed
+func sanitizeProjectedVolumesAndMounts(podSpec apiv1.PodSpec) apiv1.PodSpec {
```

Comment:
I updated this method to append a local list of volumes and set that instead of setting projected volumes to empty structs.

The reason is previously comparing:

```
p1Spec.Volumes = nil
p2Spec.Volumes = []Volume[{}] // where {} was a previous projected volume that got set to an empty struct
```
would be false.


Here's the diff:
```
 func sanitizeProjectedVolumesAndMounts(podSpec apiv1.PodSpec) apiv1.PodSpec {
-       var projectedVolumeNames []string
-       for i, v := range podSpec.Volumes {
-               if v.Projected != nil {
-                       projectedVolumeNames = append(projectedVolumeNames, v.Name)
-                       podSpec.Volumes[i] = apiv1.Volume{}
+       projectedVolumeNames := map[string]bool{}
+       var volumes []apiv1.Volume
+       for _, v := range podSpec.Volumes {
+               if v.Projected == nil {
+                       volumes = append(volumes, v)
+               } else {
+                       projectedVolumeNames[v.Name] = true
                }
        }
+       podSpec.Volumes = volumes

-       for _, pv := range projectedVolumeNames {
-               for c := range podSpec.Containers {
-                       for v, mount := range podSpec.Containers[c].VolumeMounts {
-                               if mount.Name == pv {
-                                       podSpec.Containers[c].VolumeMounts[v] = apiv1.VolumeMount{}
-                               }
+       for i := range podSpec.Containers {
+               var volumeMounts []apiv1.VolumeMount
+               for _, mount := range podSpec.Containers[i].VolumeMounts {
+                       if ok := projectedVolumeNames[mount.Name]; !ok {
+                               volumeMounts = append(volumeMounts, mount)
                        }
                }
+               podSpec.Containers[i].VolumeMounts = volumeMounts
        }
-
        return podSpec
```